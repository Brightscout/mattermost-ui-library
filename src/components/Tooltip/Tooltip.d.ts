import {OverlayTriggerProps} from 'react-bootstrap';

import {IconType} from '@Components/Icon';

/**
 * Types for position
 *
 * `auto` - renders on the side where more space is available
 *
 * `top` - renders on the top of the wrapped component
 *
 * `right` - renders on the right of the wrapped component
 *
 * `bottom` - renders on the bottom of the wrapped component
 *
 * `left` - renders on the left of the wrapped component
 */
type placementType = 'auto' | 'top' | 'right' | 'bottom' | 'left';

export interface TooltipProps extends Pick<OverlayTriggerProps, 'children'> {

    /**
	 * Placement of the tooltip with respect to the wrapped component
	 */
    placement: placementType;

    /**
	 * Text to be displayed in the tooltip
	 */
    text: string;

    /**
	 * Low-emphasis text which adds more details
	 */
    hint?: string;

    /**
	 * Icon to be displayed in the tooltip
	 */
    iconName?: Exclude<IconType, 'Spinner'>;
}
