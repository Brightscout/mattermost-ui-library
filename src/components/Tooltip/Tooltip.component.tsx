import React from 'react';
import {OverlayTrigger} from 'react-bootstrap';
import {OverlayInjectedProps} from 'react-bootstrap/esm/Overlay';

import {Icon} from '@Components/Icon';

import {TooltipProps} from './Tooltip';
import {Hint, Text, StyledTooltip} from './Tooltip.styles';

/**
 * Tooltip Component
 *
 * @example Correct usage
 *
 * ```ts
 * <Tooltip text='text'>
 *   <SomeComponent/>
 * </Tooltip>
 * ```
 * @example usage with icon and hint text
 *
 * ```ts
 * <Tooltip text='text' hint='hint text' iconName='Edit'>
 *   <SomeComponent/>
 * </Tooltip>
 * ```
 */
export const Tooltip = (props: TooltipProps) => {
    const {children, placement, iconName, text, hint} = props;

    const renderTooltip = (overlayProps: OverlayInjectedProps) => (
        <StyledTooltip
            id='tooltip'
            {...overlayProps}
        >
            <Text>
                {iconName &&
                <Icon
                    name={iconName}
                    size={12}
                />}
                {text}
            </Text>
            {hint && <Hint>{hint}</Hint>}
        </StyledTooltip>
    );

    return (
        <OverlayTrigger
            key={placement}
            placement={placement}
            flip={true}
            delay={400}
            overlay={renderTooltip}
        >
            {children}
        </OverlayTrigger>
    );
};
