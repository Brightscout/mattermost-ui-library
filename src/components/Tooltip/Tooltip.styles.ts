import {Tooltip} from 'react-bootstrap';
import styled from 'styled-components';

import colors from '@Styles/colorsForJs.module.scss';

// styling for styled tooltip
export const StyledTooltip = styled(Tooltip)({
    '.tooltip-inner': {
        background: colors.tooltipBg,
        border: 'none',
    },

    // styling for tooltip with placement top
    '&[x-placement="top"] .tooltip-arrow::before': {
        borderTopColor: colors.tooltipBg,
        transform: 'translateY(1px)',
    },

    // styling for tooltip with placement bottom
    '&[x-placement="bottom"] .tooltip-arrow::before': {
        borderBottomColor: colors.tooltipBg,
        transform: 'translateY(-1px)',
    },

    // styling for tooltip with placement right
    '&[x-placement="right"] .tooltip-arrow::before': {
        borderRightColor: colors.tooltipBg,
        transform: 'translateX(-1px)',
    },

    // styling for tooltip with placement left
    '&[x-placement="left"] .tooltip-arrow::before': {
        borderLeftColor: colors.tooltipBg,
        transform: 'translateX(1px)',
    },

    svg: {
        display: 'block',
    },
});

export const Text = styled.div({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    gap: 6,
    fontSize: 12,
});

export const Hint = styled.div({
    fontSize: 10,
    color: colors.tooltipHint,
});
