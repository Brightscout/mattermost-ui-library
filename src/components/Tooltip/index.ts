export {Tooltip} from './Tooltip.component';
export type {TooltipProps} from './Tooltip';
