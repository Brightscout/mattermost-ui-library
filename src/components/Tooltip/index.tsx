import React from 'react';
import {Placement} from 'react-bootstrap/esm/types';
import {OverlayTrigger, Tooltip as ReactTooltip} from 'react-bootstrap';

// Constants
import {TooltipDefaultPlacement} from '@Constants';

type TooltipProps = {
    tooltipContent: JSX.Element | string
    children: JSX.Element
    placement?: Placement
    className?: string;
}

const Tooltip = ({
    tooltipContent,
    children,
    placement = TooltipDefaultPlacement,
    className,
}: TooltipProps): JSX.Element => (
    <OverlayTrigger
        placement={placement}
        overlay={
            <ReactTooltip
                className={`tooltip ${className}`}
                placement={placement}
            >
                {tooltipContent}
            </ReactTooltip>
        }
    >
        {children}
    </OverlayTrigger>
);

export default Tooltip;
