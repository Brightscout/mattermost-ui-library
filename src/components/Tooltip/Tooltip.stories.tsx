import React from 'react';
import {Story, Meta} from '@storybook/react';

import {Tooltip} from './Tooltip.component';
import {TooltipProps} from './Tooltip';

export default {
    title: 'Component/Tooltip',
    component: Tooltip,
    parameters: {
        layout: 'centered',
    },
} as Meta<typeof Tooltip>;

const TooltipTemplate: Story<TooltipProps> = (args) => <Tooltip {...args}/>;

// Default
export const Default = TooltipTemplate.bind({});
Default.args = {
    children: <a href='#'>{'Click here'}</a>,
    placement: 'top',
    text: 'Edit',
};

// Tooltip with hint
export const TooltipWithHintAndPlacementRight = TooltipTemplate.bind({});
TooltipWithHintAndPlacementRight.args = {
    children: <a href='#'>{'Click here'}</a>,
    placement: 'right',
    text: 'Edit',
    hint: 'edit to change the text of the link',
};

// Tooltip with hint and icon
export const TooltipWithIconAndHint = TooltipTemplate.bind({});
TooltipWithIconAndHint.args = {
    children: <a href='#'>{'Click here'}</a>,
    placement: 'top',
    text: 'Edit',
    hint: 'edit to change the text of the link',
    iconName: 'Edit',
};
