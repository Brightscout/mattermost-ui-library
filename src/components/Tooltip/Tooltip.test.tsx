import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';

// Mock data
import {
    TooltipChildren,
    TooltipContent,
    TooltipClassName
} from '@Constants/mockData/tooltip';

import Tooltip from '.';

describe('Tooltip', () => {
    let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

    beforeEach(() => {
        component = shallow(
            <Tooltip
                className={TooltipClassName}
                tooltipContent={TooltipContent}
            >
                {TooltipChildren}
            </Tooltip>,
        );
    });

    it('Should render correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('Should render the children correctly', () => {
        expect(component.contains(TooltipChildren)).toBeTruthy();
    });
});
