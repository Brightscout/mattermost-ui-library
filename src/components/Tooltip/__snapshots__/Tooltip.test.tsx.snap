// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Should render correctly 1`] = `
<OverlayTrigger
  defaultShow={false}
  overlay={
    <ForwardRef
      className="tooltip tooltip-class"
      placement="top"
    >
      <span
        className="tooltip-content-class"
      >
        Tooltip content
      </span>
    </ForwardRef>
  }
  placement="top"
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <span
    className="tooltip-children-class"
  >
    Tooltip Children
  </span>
</OverlayTrigger>
`;
